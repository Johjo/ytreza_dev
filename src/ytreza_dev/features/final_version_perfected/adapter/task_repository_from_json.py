import json
from dataclasses import asdict
from pathlib import Path
from typing import Any

from ytreza_dev.features.final_version_perfected.port.task_repository import FvpRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, TaskNew, TaskNext, TaskLater, TaskNever


class FvpRepositoryFromJson(FvpRepositoryPort):
    def all_tasks(self) -> list[TaskBase]:

        tasks = json.loads(self._file_path.read_text())
        return [self.to_task_base(task) for task in tasks]

    @staticmethod
    def to_task_base(task: dict[str, str]) -> TaskBase:
        match task["status"]:
            case "new":
                return TaskNew(id=task["id"])
            case "next":
                return TaskNext(id=task["id"])
            case "later":
                return TaskLater(id=task["id"])
            case "never":
                return TaskNever(id=task["id"])
        raise ValueError(f"Unknown status: {task['status']}")

    def __init__(self, file_path: Path) -> None:
        self._file_path = file_path

    def save(self, tasks: list[TaskBase]) -> None:
        tasks_dict = [self._to_dict(task) for task in tasks]
        self._file_path.write_text(json.dumps(tasks_dict, indent=4))

    @staticmethod
    def _to_dict(task: TaskBase) -> dict[str, Any]:
        d = asdict(task)
        match task:
            case TaskNext():
                d.update({"status": "next"})
            case TaskNew():
                d.update({"status": "new"})
            case TaskLater():
                d.update({"status": "later"})
            case TaskNever():
                d.update({"status": "never"})

        return d