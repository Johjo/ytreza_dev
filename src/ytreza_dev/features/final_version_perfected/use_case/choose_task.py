from ytreza_dev.features.final_version_perfected.port.task_repository import TaskRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, TaskNew


class ChooseTaskUseCase:
    def __init__(self, task_repository: TaskRepositoryPort):
        self._task_repository = task_repository

    def execute(self, url: str) -> None:
        all_tasks_before = self._task_repository.all_tasks()
        all_tasks_after : list[TaskBase] = []

        all_tasks_after.append(all_tasks_before[0].to_next())

        first_new_index = self._first_new_index(all_tasks_before[1:]) + 1
        i = 1

        while len(all_tasks_before) > i:
            if url == all_tasks_before[i].url:
                all_tasks_after.append(all_tasks_before[i].to_next())
            else:
                if i != first_new_index:
                    all_tasks_after.append(all_tasks_before[i])
                else:
                    all_tasks_after.append(all_tasks_before[i].to_later())
            i += 1

        self._task_repository.save(all_tasks_after)

    @staticmethod
    def _first_new_index(all_tasks_before: list[TaskBase]) -> int:
        for index, task in enumerate(all_tasks_before):
            if isinstance(task, TaskNew):
                return index

        raise NotImplementedError()
