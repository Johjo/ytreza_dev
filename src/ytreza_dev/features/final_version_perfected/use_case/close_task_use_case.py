from ytreza_dev.features.final_version_perfected.port.external_todolist import ExternalTodolistPort
from ytreza_dev.features.final_version_perfected.port.task_repository import TaskFvpRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskNever, TaskBase


class CloseTaskUseCase:
    def __init__(self, task_repository: TaskFvpRepositoryPort, external_todolist: ExternalTodolistPort) -> None:
        self._external_todolist = external_todolist
        self._task_repository = task_repository

    def execute(self, url: str) -> None:
        after, closed_task, before = self._extract_before_and_task_and_after(self._task_repository.all_tasks(), url)
        after += self._update_task_following_closed_task(before)
        if all(isinstance(task, TaskNever) for task in after):
            after = [task.to_new() for task in after]
        self._task_repository.save(after)
        self._external_todolist.close_task(url=url, task_id=closed_task.id)

    @staticmethod
    def _update_task_following_closed_task(before: list[TaskBase]) -> list[TaskBase]:
        after : list[TaskBase] = []
        for task in before:
            if isinstance(task, TaskNever):
                after.append(task)
            else:
                after.append(task.to_new())
        return after

    @staticmethod
    def _extract_before_and_task_and_after(before: list[TaskBase], url: str) -> tuple[list[TaskBase], TaskBase,list[TaskBase]]:
        after = []
        # on garde toutes les tâches avant la tâche à fermer
        i = 0
        while before[i].url != url:
            after.append(before[i])
            i += 1

        closed_task = before[i]
        # on ignore la tâche fermée
        before = before[i + 1:]
        return after, closed_task, before
