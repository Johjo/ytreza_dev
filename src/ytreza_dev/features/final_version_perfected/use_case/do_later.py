from ytreza_dev.features.final_version_perfected.port.task_repository import TaskFvpRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskBase


class DoLater:
    def __init__(self, task_repository: TaskFvpRepositoryPort) -> None:
        self._task_repository = task_repository

    def execute(self, updated_key: str) -> None:
        before : list[TaskBase] = self._task_repository.all_tasks()

        self._task_repository.save([before[0].to_next()] + [self._update_task(task, updated_key) for task in before[1:]])

    @staticmethod
    def _update_task(task: TaskBase, updated_key: str) -> TaskBase:
        if task.id != updated_key:
            return task

        return task.to_later()
