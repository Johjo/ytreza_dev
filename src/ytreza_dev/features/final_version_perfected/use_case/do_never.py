from ytreza_dev.features.final_version_perfected.types import TaskBase


class DoNever:
    def __init__(self, task_repository):
        self._task_repository = task_repository

    def execute(self, url: str):
        before : list[TaskBase] = self._task_repository.all_tasks()

        self._task_repository.save([before[0].to_next()] + [self._update_task(task, url) for task in before[1:]])

    @staticmethod
    def _update_task(task: TaskBase, url):
        if task.url != url:
            return task

        return task.to_never()
