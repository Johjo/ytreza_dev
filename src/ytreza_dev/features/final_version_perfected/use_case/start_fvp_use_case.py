from expression import Nothing

from ytreza_dev.features.final_version_perfected.port.task_information_repository import TaskInformationRepositoryPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.port.task_repository import FvpRepositoryPort
from ytreza_dev.features.final_version_perfected.port.todolist_reader import TodolistReaderPort
from ytreza_dev.features.final_version_perfected.types import TaskNew, ExternalTask, ExternalProject, Project


class StartFvpUseCase:
    def __init__(self, todolist_reader: TodolistReaderPort, fvp_repository: FvpRepositoryPort, task_information_repository: TaskInformationRepositoryPort) -> None:
        self._task_information_repository = task_information_repository
        self._todolist_reader = todolist_reader
        self._fvp_repository = fvp_repository

    def execute(self) -> None:

        all_external_tasks : list[ExternalTask] = self._todolist_reader.all_tasks()
        self._fvp_repository.save([self._to_task_new(task) for task in all_external_tasks])
        self._task_information_repository.save([self._to_task_information(task) for task in all_external_tasks])

    @staticmethod
    def _to_task_information(external_task: ExternalTask) -> TaskInformation:
        return TaskInformation(
            key=external_task.id,
            title=external_task.name,
            project=Project(key=external_task.project.key, name=external_task.project.name),
            url=external_task.url,
            due_date=Nothing
        )

    @staticmethod
    def _to_task_new(task: ExternalTask) -> TaskNew :
        return TaskNew(id=task.id)

    @staticmethod
    def _to_project(project: ExternalProject) -> Project:
        return Project(key=project.key, name=project.name)
