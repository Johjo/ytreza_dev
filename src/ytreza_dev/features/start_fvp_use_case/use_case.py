from abc import ABCMeta, abstractmethod
from dataclasses import dataclass

from ytreza_dev.shared.final_version_perfected.types import TaskBase, TaskNew


@dataclass
class OldTask:
    name: str
    url: str


class TaskRepository(metaclass=ABCMeta):
    @abstractmethod
    def save(self, tasks: list[TaskBase]) -> None:
        pass

@dataclass
class ExternalTask:
    name: str
    url: str


class TodolistReader(metaclass=ABCMeta):
    @abstractmethod
    def all_tasks(self) -> list[ExternalTask]:
        pass


class StartFvpUseCase:
    def __init__(self, todolist_reader: TodolistReader, task_repository: TaskRepository) -> None:
        self._todolist_reader = todolist_reader
        self._task_repository = task_repository

    def execute(self) -> None:
        self._task_repository.save([TaskNew(title=task.name, url=task.url) for task in self._todolist_reader.all_tasks()])
