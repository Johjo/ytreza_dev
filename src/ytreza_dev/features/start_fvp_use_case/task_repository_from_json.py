import json
from dataclasses import asdict
from pathlib import Path

from ytreza_dev.features.start_fvp_use_case.use_case import TaskRepository, OldTask
from ytreza_dev.shared.final_version_perfected.types import TaskBase


class TaskRepositoryFromJson(TaskRepository):
    def all_tasks(self) -> list[TaskBase]:
        raise NotImplementedError()

    def __init__(self, file_path: Path) -> None:
        self._file_path = file_path

    def save(self, tasks: list[TaskBase]) -> None:
        tasks_dict = [self._to_dict(task) for task in tasks]
        self._file_path.write_text(json.dumps(tasks_dict, indent=4))

    @staticmethod
    def _to_dict(task: TaskBase):
        d = asdict(task)
        d.update({"status": "new"})
        return d
