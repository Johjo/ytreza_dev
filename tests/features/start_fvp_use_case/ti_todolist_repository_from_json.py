from approvaltests import verify
from pathlib import Path

from ytreza_dev.features.start_fvp_use_case.task_repository_from_json import TaskRepositoryFromJson
from ytreza_dev.features.start_fvp_use_case.use_case import Task


def test_write_in_json():
    json_path = Path("./data_test/tasks.json")
    sut = TaskRepositoryFromJson(json_path)

    sut.save([
        Task(name="buy the milk", url="https://url_1.com"),
        Task(name="buy the water", url="https://url_2.com")])

    verify(Path("tasks.json").read_text(encoding="utf-8"))

