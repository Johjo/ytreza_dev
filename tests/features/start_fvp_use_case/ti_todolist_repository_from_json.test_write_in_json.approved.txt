[
    {
        "name": "buy the milk",
        "url": "https://url_1.com",
        "status": "new"
    },
    {
        "name": "buy the water",
        "url": "https://url_2.com",
        "status": "new"
    }
]
